public/app/plugins/datasource/cloud-monitoring/components/Experimental/Metrics.tsx

```diff
@@ -0,0 +1,184 @@
+import { css } from '@emotion/css';
+import { startCase, uniqBy } from 'lodash';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+import { EditorRow, EditorField, EditorFieldGroup } from '@grafana/experimental';
+import { getSelectStyles, Select, useStyles2, useTheme2 } from '@grafana/ui';
+
+import CloudMonitoringDatasource from '../../datasource';
+import { MetricDescriptor } from '../../types';
+
+export interface Props {
+  refId: string;
+  onChange: (metricDescriptor: MetricDescriptor) => void;
+  templateVariableOptions: Array<SelectableValue<string>>;
+  datasource: CloudMonitoringDatasource;
+  projectName: string;
+  metricType: string;
+  children: (metricDescriptor?: MetricDescriptor) => JSX.Element;
+}
+
+interface State {
+  metricDescriptors: MetricDescriptor[];
+  metrics: any[];
+  services: any[];
+  service: string;
+  metric: string;
+  metricDescriptor?: MetricDescriptor;
+  projectName: string | null;
+}
+
+export function Metrics(props: Props) {
+  const [state, setState] = useState<State>({
+    metricDescriptors: [],
```

Comment:
can you split each property on its own?  that way you don't need to update the whole state just for one property.